### PR TITLE
 KP-548/KP-644 Add method to create new cluster

### PR DESCRIPTION
**Pull request by xkcm**
* Added **Create New Cluster** method

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -43,7 +40,79 @@ async function getToken({ params }, settings) {
...
+        },
+      ],
+    };
+  }
+  if (!isObjectEmpty(tags)) {
+    clusterParams.tags = tags;
+  }
+  if (encryption.resources && !isObjectEmpty(encryption.provider)) {
+    clusterParams.encryptionConfig = [encryption];
+  }
```
Comment (kdebowski-kaholo) [inline app.js]: Over 60 lines of code in the method for data handling and 2-3 lines of the actual logic seems pretty excessive. Please consider creating payload-functions.js file and extracting the data manipulation code from this method to separate function (or ideally multiple functions, logically separated) and then replacing the 60+ lines in here with one function call.

File: autocomplete.js [hunk 2]
```diff
@@ -48,6 +50,31 @@ async function listRegions(query, pluginSettings, actionParams) {
...
+      let policy;
+      try {
+        policy = JSON.parse(decodeURIComponent(AssumeRolePolicyDocument));
+      } catch {
+        throw new Error(CORRUPTED_ROLE_POLICY_TEXT(RoleName));
+      }
+      const roleIntendedForEks = policy.Statement.some(
+        (item) => item.Principal.Service.includes(EKS_SERVICE_URL),
+      );
+      return RoleName.includes(query) && roleIntendedForEks;
```
Comment (kdebowski-kaholo) [inline autocomplete.js]: That's pretty long (and logic infused) body for a `filter` callback. Please consider extracting validation code to a separate function in helpers.js and then applying this function in `filter` instead.

**Review by kdebowski-kaholo — CHANGES_REQUESTED** [review]

**Review by kdebowski-kaholo — DISMISSED** [review]

File: consts.js [hunk 3]
```diff
@@ -1,9 +1,20 @@
 const EXPIRES_INVALID_MESSAGE = "\"Expires\" param must be an integer";
 const CLUSTER_REQUIRED_MESSAGE = "Cluster name is required";
 const MISSING_OR_INCORRECT_CREDENTIALS_ERROR = "Missing or incorrect credentials - please select valid access and secret keys first";
+const ARN_REQUIRED_MESSAGE = "ARN is required.";
+const VPC_SUBNETS_REQUIRED_MESSAGE = "VPC Subnet IDs are required. Please specify at least two.";
+const VPC_SECURITY_GROUPS_REQUIRED_MESSAGE = "VPC Security Group IDs are required.";
+const CORRUPTED_ROLE_POLICY_TEXT = (roleName) => `Failed to parse role policy for role "${roleName}".`;
```
Comment (graddaniel) [inline consts.js]: This is not a text, but a function returning the text.

File: consts.js [hunk 4]
```diff
@@ -1,9 +1,20 @@
 const EXPIRES_INVALID_MESSAGE = "\"Expires\" param must be an integer";
 const CLUSTER_REQUIRED_MESSAGE = "Cluster name is required";
 const MISSING_OR_INCORRECT_CREDENTIALS_ERROR = "Missing or incorrect credentials - please select valid access and secret keys first";
+const ARN_REQUIRED_MESSAGE = "ARN is required.";
```
Comment (graddaniel) [inline consts.js]: Some of those values are titled `ERROR` some are called `MESSAGE` while in the end they are `ERROR_MESSAGE`s. It should be refactored.

File: config.json [hunk 5]
```diff
@@ -5,8 +5,8 @@
   "imgUrl": "EKS.png",
   "execProgram": "node",
   "main": "app.js",
-  "version": "1.1.0",
-  "description": "Works with AWS EKS API",
+  "version": "2.0.0",
```
Comment (graddaniel) [inline config.json]: You're introducing only a new method, nothing that breaks the backwards compatibility - no need to update the major version. It should be `1.2.0`.

File: app.js [hunk 6]
```diff
@@ -1,30 +1,20 @@
...
+} = require("./payload-functions");
 
 async function getToken({ params }, settings) {
-  const clusterName = parsers.string(params.clusterName);
-  const region = parsers.autocomplete(params.region);
-  const expires = parsers.integer(params.expires) || 60;
-  const secretAccessKey = parsers.string(params.secretAccessKey || settings.secretAccessKey);
-  const accessKeyId = parsers.string(params.accessKeyId || settings.accessKeyId);
+  const payload = createPayloadForGetToken(params);
+  validateGetTokenPayload(payload);
```
Comment (graddaniel) [inline app.js]: You're creating a `payload` and then you're checking if it's been done correctly with `validate` function. Don't you trust your own code? ;)
If some of the `payload` fields rely on data which could be missing, then put the checks there, to make sure they exist before you use it.

Also, considering that we have configuration fields allowing the platform to verify those, there's no need to put that in the code and duplicate the functionality. Please set the required constraints in the `config.json` file.
Paragraph 3 of the following document describes how to achieve that:
https://docs.google.com/document/d/1e7wVLj4_667C8QiztDUXePiW6faffb33unjxWYUTHAE/edit#heading=h.r2jqx4eb8u7x
Let me know if there are problematic ones.

File: app.js [hunk 7]
```diff
@@ -33,17 +23,28 @@ async function getToken({ params }, settings) {
...
     token,
     clusterHost: cluster.endpoint,
     clusterCA: cluster.certificateAuthority.data,
   };
 }
 
+async function createCluster({ params }, settings) {
+  const { accessKeyId, secretAccessKey, region } = mapAWSConfig(params, settings);
+  const clusterPayload = createPayloadForCreateCluster(params);
+  validateCreateClusterPayload(clusterPayload);
```
Comment (graddaniel) [inline app.js]: same here

**Review by graddaniel — CHANGES_REQUESTED** [review]

File: consts.js [hunk 8]
```diff
@@ -1,9 +1,11 @@
-const EXPIRES_INVALID_MESSAGE = "\"Expires\" param must be an integer";
-const CLUSTER_REQUIRED_MESSAGE = "Cluster name is required";
-const MISSING_OR_INCORRECT_CREDENTIALS_ERROR = "Missing or incorrect credentials - please select valid access and secret keys first";
+const MISSING_OR_INCORRECT_CREDENTIALS_MESSAGE = "Missing or incorrect credentials - please select valid access and secret keys first";
+
+const EKS_SERVICE_URL = "eks.amazonaws.com";
+
+const DEFAULT_TOKEN_EXPIRE_TIME = 60;
```
Comment (graddaniel) [inline consts.js]: Why don't you specify the value via `default` property in the `config.json`?

File: consts.js [hunk 9]
```diff
@@ -1,9 +1,11 @@
-const EXPIRES_INVALID_MESSAGE = "\"Expires\" param must be an integer";
-const CLUSTER_REQUIRED_MESSAGE = "Cluster name is required";
-const MISSING_OR_INCORRECT_CREDENTIALS_ERROR = "Missing or incorrect credentials - please select valid access and secret keys first";
+const MISSING_OR_INCORRECT_CREDENTIALS_MESSAGE = "Missing or incorrect credentials - please select valid access and secret keys first";
```
Comment (graddaniel) [inline consts.js]: This message doesn't seem to be used anywehere.

File: helpers.js [hunk 10]
```diff
@@ -1,23 +1,52 @@
...
+const { EKS_SERVICE_URL } = require("./consts");
 const parsers = require("./parsers");
 
-function getEc2(params, settings) {
-  return new aws.EC2({
-    region: parsers.autocomplete(params.region) || settings.region,
-    accessKeyId: params.accessKeyId || settings.accessKeyId,
-    secretAccessKey: params.secretAccessKey || settings.secretAccessKey,
-  });
+const createCorruptedRolePolicyMessage = (roleName) => `Failed to parse role policy for role "${roleName}".`;
```
Comment (graddaniel) [inline helpers.js]: Name looks good, but do we really need to wrap a template with a function?
It's not reused anywhere, and it doesn't simplify the code, thus there's no benefit (unless there's something I'm missing).

I would just pass the string template directly to the `Error` object constructor.

In future we'll think of creating a common set of reusable errors shared with `utils` (or something similar) library, which will hide such details.

File: app.js [hunk 11]
```diff
@@ -33,17 +21,27 @@ async function getToken({ params }, settings) {
...
+    expirationTimestamp,
     token,
     clusterHost: cluster.endpoint,
     clusterCA: cluster.certificateAuthority.data,
   };
 }
 
+async function createCluster({ params }, settings) {
+  const { accessKeyId, secretAccessKey, region } = mapAwsConfig(params, settings);
+  const clusterPayload = createPayloadForCreateCluster(params);
```
Comment (graddaniel) [inline app.js]:
```suggestion
  const clusterPayload = createPayloadForCreateCluster(params);
  
  const { accessKeyId, secretAccessKey, region } = mapAwsConfig(params, settings);
```
Tiny reordering emphasizing that there are other similar blocks in other functions, which might give a hint for some further refactoring.

**Review by graddaniel — CHANGES_REQUESTED** [review]
Looks better, yet few things still require improving.

File: consts.js (hunk 9 above)
Comment (xkcm) [reply consts.js]: It is used in [autocomplete.js](https://github.com/Kaholo/kaholo-plugin-aws-eks/pull/5/files#diff-8d2a05e5cc6543b2285d1de69039fa5a6c29757286c71d9ff81f68aafab30886) at lines [58](https://github.com/Kaholo/kaholo-plugin-aws-eks/pull/5/files#diff-8d2a05e5cc6543b2285d1de69039fa5a6c29757286c71d9ff81f68aafab30886R58) and [49](https://github.com/Kaholo/kaholo-plugin-aws-eks/pull/5/files#diff-8d2a05e5cc6543b2285d1de69039fa5a6c29757286c71d9ff81f68aafab30886R49).

Comment (graddaniel) [reply consts.js]: It's `MISSING_OR_INCORRECT_CREDENTIALS_ERROR`. Running this code would throw type errors.
Please rename it to `MISSING_OR_INCORRECT_CREDENTIALS_ERROR_MESSAGE` to emphasize it's function.

**Review by graddaniel — APPROVED** [review]
GJ!

**Review by kdebowski-kaholo — APPROVED** [review]
Looks much better, good job!

**Pull request merged**